### ISSUE TITLE
SixLabors.ImageSharp.ImageProcessingException

### ISSUE THREAD

**Issue by zemien** [johannesegger/SenseHatNet#2]
I'm using the latest .Net Core 2.1 (2.1.300) and created a new console app that has 2 commands:

- Console.WriteLine("Hello World");
- Sense.Led.LedMatrix.ShowMessage("Hello World");

Published and copied it over to Raspbian Jessie. Executed it in shell. The following happened:
- The console showed Hello World
- The SenseHat LEDs turned off
- I get the exception `SixLabors.ImageSharp.ImageProcessingException: An error occured when processing the image using FillRegionProcessor``1`

Any ideas why this might be happening? I expected the NuGet package would pull down any other dependencies it needed.

**Comment by johannesegger** [comment]
Thanks for reporting this.  
To be honest I have no idea why this might happen.  
Can you try using the latest version of `SixLabors.ImageSharp` (`dotnet add package SixLabors.ImageSharp`)?  
Can you also try running the sample app on your development machine? Do you get the same exception (doesn't happen on my machine) or do you get an exception about some missing file (that would be expected because it's some Sense HAT specific file)?

**Comment by zemien** [comment]
Yes I get the same exception on development machine (Mac). 
```
Unhandled Exception: SixLabors.ImageSharp.ImageProcessingException: An error occured when processing the image using FillRegionProcessor`1. See the inner exception for more detail. ---> 
SixLabors.ImageSharp.ImageProcessingException: An error occured when processing the image using FillRegionProcessor`1. See the inner exception for more detail. ---> 
System.MissingMethodException: Method not found: 'SixLabors.ImageSharp.Rgba32 ByRef System.Span`1.DangerousGetPinnableReference()'.
```

I get the same exception with SixLabors.ImageSharp 1.0.0-beta001 (the minimum dependency of your package).

I then added SixLabors.ImageSharp 1.0.0-beta004 and now I get the exception
```
Unhandled Exception: System.TypeLoadException: Could not load type 'SixLabors.ImageSharp.Rgba32' from assembly 'SixLabors.ImageSharp, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'.
```
on both Raspberry Pi and Mac.

**Comment by johannesegger** [comment]
Ohh sorry, I meant `dotnet add SixLabors.ImageSharp.Drawing --version 1.0.0-beta0004`.  
Alternatively you could try cloning this repository and build/run `SenseTest/SenseTest.csproj`.

**Comment by zemien** [comment]
I get the same exception as my initial post if I just add `dotnet add SixLabors.ImageSharp.Drawing --version 1.0.0-beta0004`

I looked into the SenseTest file and extracted the test for the sensors. I get results back which at least suggests the physical connection and driver is working:

![image](https://user-images.githubusercontent.com/8407032/41646385-bf47ea4c-74c7-11e8-81fe-83a3ba0c3a03.png)

As a side question, what do the valid flags mean? I would love to contribute some wiki to this project.

**Comment by johannesegger** [comment]
Because you get the exception when running this on your development machine I think it's got to do with `SixLabors.ImageSharp` only or with the way we use it. Unfortunately I still didn't manage to reproduce it (a failing test in a CI environment would be nice).

Regarding your second question: This comes from [RTIMULib](https://github.com/RPi-Distro/RTIMULib.git) and I've no idea what most of that information means to be honest.

**Comment by johannesegger** [comment]
I've updated the [sample application](https://github.com/johannesegger/SenseHatNet/tree/master/Sample) because .NET Core 2.1 now allows building the app directly on a RPi. If you have installed Docker on your Pi you could try running [build.sh](https://github.com/johannesegger/SenseHatNet/blob/master/Sample/build.sh) which creates a Docker container out of the sample app and then [run.sh](https://github.com/johannesegger/SenseHatNet/blob/master/Sample/run.sh) to run the app. I ran this on a remotely located RPi 3 and I didn't get an exception. On Monday I can check if the sample app really draws something on the LED matrix.

**Comment by zemien** [comment]
Thanks! I built and ran the sample app on my Pi and indeed I see the Hello World flash across the LEDs. So that proves that SenseHatNet library and my hardware is fine. I wonder if it has anything to do with compiling my app on Mac OS?

**Comment by johannesegger** [comment]
You could try running [PixelsFromText.Create](https://github.com/johannesegger/SenseHatNet/blob/master/Sense/Led/PixelsFromText.cs#L13) directly. This has nothing to do with Sense HAT and should work on any other machine as well.

**Comment by johannesegger** [comment]
I played a bit more with this and got the same exception (see below) when I removed `-r linux-arm` from `dotnet publish`. So it only works when I build the app using `dotnet publish -r linux-arm`.

```
Unhandled Exception: SixLabors.ImageSharp.ImageProcessingException: An error occured when processing the image using FillRegionProcessor`1. See the inner exception for more detail. ---> SixLabors.ImageSharp.ImageProcessingException: An error occured when processing the image using FillRegionProcessor`1. See the inner exception for more detail. ---> System.MissingMethodException: Method not found: 'SixLabors.ImageSharp.Rgba32 ByRef System.Span`1.DangerousGetPinnableReference()'.
   at SixLabors.ImageSharp.PixelFormats.PixelOperations`1.ToVector4(Span`1 sourceColors, Span`1 destVectors, Int32 count)
   at SixLabors.ImageSharp.Rgba32.PixelOperations.ToVector4(Span`1 sourceColors, Span`1 destVectors, Int32 count)
   at SixLabors.ImageSharp.PixelFormats.PixelBlenders.DefaultPixelBlenders`1.Normal.Blend(Span`1 destination, Span`1 background, Span`1 source, Span`1 amount)
   at SixLabors.ImageSharp.Drawing.Brushes.SolidBrush`1.SolidBrushApplicator.Apply(Span`1 scanline, Int32 x, Int32 y)
   at SixLabors.ImageSharp.Drawing.Processors.FillRegionProcessor`1.OnApply(ImageFrame`1 source, Rectangle sourceRectangle, Configuration configuration)
   at SixLabors.ImageSharp.Processing.ImageProcessor`1.Apply(ImageFrame`1 source, Rectangle sourceRectangle, Configuration configuration)
   --- End of inner exception stack trace ---
   at SixLabors.ImageSharp.Processing.ImageProcessor`1.Apply(ImageFrame`1 source, Rectangle sourceRectangle, Configuration configuration)
   at SixLabors.ImageSharp.Processing.ImageProcessor`1.Apply(Image`1 source, Rectangle sourceRectangle)
   --- End of inner exception stack trace ---
   at SixLabors.ImageSharp.Processing.ImageProcessor`1.Apply(Image`1 source, Rectangle sourceRectangle)
   at SixLabors.ImageSharp.DefaultInternalImageProcessorContext`1.ApplyProcessor(IImageProcessor`1 processor, Rectangle rectangle)
   at SixLabors.ImageSharp.DefaultInternalImageProcessorContext`1.ApplyProcessor(IImageProcessor`1 processor)
   at SixLabors.ImageSharp.ImageExtensions.Fill[TPixel](IImageProcessingContext`1 source, IBrush`1 brush, IPathCollection paths, GraphicsOptions options)
   at SixLabors.ImageSharp.ImageExtensions.DrawText[TPixel](IImageProcessingContext`1 source, String text, Font font, IBrush`1 brush, IPen`1 pen, PointF location, TextGraphicsOptions options)
   at SixLabors.ImageSharp.ImageExtensions.DrawText[TPixel](IImageProcessingContext`1 source, String text, Font font, TPixel color, PointF location)
   at Sense.Led.PixelsFromText.<>c__DisplayClass0_1.<Create>b__0(IImageProcessingContext`1 ctx)
   at SixLabors.ImageSharp.ImageExtensions.Mutate[TPixel](Image`1 source, Action`1 operation)
   at Sense.Led.PixelsFromText.Create(String text)
   at Sense.Led.LedMatrix.ShowMessage(String text)
   at SenseTest.Program.TestLedMessage(String text) in /usr/src/Program.cs:line 24
   at SenseTest.Program.Main(String[] args) in /usr/src/Program.cs:line 16
```

**Comment by tgabee2** [comment]
Does anybody have a working solution with .Net Core 2.1 on Linux (Ubuntu Mate)? I tried all the idea above, but not working, same exception is thrown. I published to linux-arm as well, but no luck.
Adding the latest SixLabors.ImageSharp package creates a different error, Rgba32 type is not found (I think it's moved to a different assembly in the latest package).
Any suggestion?

**Comment by johannesegger** [comment]
Sorry for the delay - I don't always have my RPi with me.

When I'm running a .Net Core 2.1 docker container on my RPi I don't get that exception. Can you try installing docker on your RPi and then running the [sample app](https://github.com/johannesegger/SenseHatNet/tree/master/Sample) using [build.sh](https://github.com/johannesegger/SenseHatNet/blob/master/Sample/build.sh) and [run.sh](https://github.com/johannesegger/SenseHatNet/blob/master/Sample/run.sh) (both invoked from your RPi)?

**Comment by johannesegger** [comment]
I'm closing this for now, please re-open if need be.

**Comment by tgabee2** [comment]
Hi,

Finally I got the chance to try the simple Hello World app with Docker, and unfortunately the result is the same:

```
gabz@gaPi:~/Projects$ docker build -t sense-hat-net .
Sending build context to Docker daemon  71.91MB
Step 1/9 : FROM microsoft/dotnet:2.1-sdk as builder
 ---> e1ca2693e8f4
Step 2/9 : WORKDIR /usr/src/
 ---> Using cache
 ---> f00d1d99e7a7
Step 3/9 : COPY Coreplay2_Local .
 ---> Using cache
 ---> 3abc6a7f92c2
Step 4/9 : RUN dotnet publish -r linux-arm -o /usr/src/dist
 ---> Using cache
 ---> 1e2ea6aaf519
Step 5/9 : FROM microsoft/dotnet:2.1-runtime
 ---> ff658372e54a
Step 6/9 : WORKDIR /usr/app/
 ---> Using cache
 ---> eeadfe0b1b44
Step 7/9 : COPY --from=builder /usr/src/dist .
 ---> Using cache
 ---> a903e4ad957f
Step 8/9 : RUN apt-get update &&     apt-get install libunwind8
 ---> Using cache
 ---> 3598429862fc
Step 9/9 : ENTRYPOINT [ "./Coreplay2_Local" ]
 ---> Running in ba80c451d054
Removing intermediate container ba80c451d054
 ---> ba89980ad627
Successfully built ba89980ad627
Successfully tagged sense-hat-net:latest
gabz@gaPi:~/Projects$ docker run --privileged --rm -v /sys/class/graphics:/sys/class/graphics -v /sys/class/input:/sys/class/input -v /dev/:/dev/ sense-hat-net
Hello World!

Unhandled Exception: SixLabors.ImageSharp.ImageProcessingException: An error occured when processing the image using FillRegionProcessor`1. See the inner exception for more detail. ---> SixLabors.ImageSharp.ImageProcessingException: An error occured when processing the image using FillRegionProcessor`1. See the inner exception for more detail. ---> System.MissingMethodException: Method not found: 'SixLabors.ImageSharp.Rgba32 ByRef System.Span`1.DangerousGetPinnableReference()'.
   at SixLabors.ImageSharp.PixelFormats.PixelOperations`1.ToVector4(Span`1 sourceColors, Span`1 destVectors, Int32 count)
   at SixLabors.ImageSharp.Rgba32.PixelOperations.ToVector4(Span`1 sourceColors, Span`1 destVectors, Int32 count)
   at SixLabors.ImageSharp.PixelFormats.PixelBlenders.DefaultPixelBlenders`1.Normal.Blend(Span`1 destination, Span`1 background, Span`1 source, Span`1 amount)
   at SixLabors.ImageSharp.Drawing.Brushes.SolidBrush`1.SolidBrushApplicator.Apply(Span`1 scanline, Int32 x, Int32 y)
   at SixLabors.ImageSharp.Drawing.Processors.FillRegionProcessor`1.OnApply(ImageFrame`1 source, Rectangle sourceRectangle, Configuration configuration)
   at SixLabors.ImageSharp.Processing.ImageProcessor`1.Apply(ImageFrame`1 source, Rectangle sourceRectangle, Configuration configuration)
   --- End of inner exception stack trace ---
   at SixLabors.ImageSharp.Processing.ImageProcessor`1.Apply(ImageFrame`1 source, Rectangle sourceRectangle, Configuration configuration)
   at SixLabors.ImageSharp.Processing.ImageProcessor`1.Apply(Image`1 source, Rectangle sourceRectangle)
   --- End of inner exception stack trace ---
   at SixLabors.ImageSharp.Processing.ImageProcessor`1.Apply(Image`1 source, Rectangle sourceRectangle)
   at SixLabors.ImageSharp.DefaultInternalImageProcessorContext`1.ApplyProcessor(IImageProcessor`1 processor, Rectangle rectangle)
   at SixLabors.ImageSharp.DefaultInternalImageProcessorContext`1.ApplyProcessor(IImageProcessor`1 processor)
   at SixLabors.ImageSharp.ImageExtensions.Fill[TPixel](IImageProcessingContext`1 source, IBrush`1 brush, IPathCollection paths, GraphicsOptions options)
   at SixLabors.ImageSharp.ImageExtensions.DrawText[TPixel](IImageProcessingContext`1 source, String text, Font font, IBrush`1 brush, IPen`1 pen, PointF location, TextGraphicsOptions options)
   at SixLabors.ImageSharp.ImageExtensions.DrawText[TPixel](IImageProcessingContext`1 source, String text, Font font, TPixel color, PointF location)
   at Sense.Led.PixelsFromText.<>c__DisplayClass0_1.<Create>b__0(IImageProcessingContext`1 ctx)
   at SixLabors.ImageSharp.ImageExtensions.Mutate[TPixel](Image`1 source, Action`1 operation)
   at Sense.Led.PixelsFromText.Create(String text)
   at Sense.Led.LedMatrix.ShowMessage(String text)
   at Coreplay2_Local.Program.Main(String[] args) in /usr/src/Program.cs:line 10
```

**Comment by johannesegger** [comment]
The error says it can't find `Span<T>.DangerousGetPinnableReference()`. According to [a comment in an `ImageSharp` issue](https://github.com/SixLabors/ImageSharp/issues/459#issuecomment-368999242) they changed the API to not use `Span<T>.DangerousGetPinnableReference()`. The current version of `SixLabors.ImageSharp` (i.e. `1.0.0-beta0005`) should have that change. So could you please try upgrading `SixLabors.ImageSharp.Drawing` to version `1.0.0-beta0005`?

**Comment by tgabee2** [comment]
I tried, the exception is changed to:

Unhandled Exception: System.TypeLoadException: Could not load type 'SixLabors.ImageSharp.Rgba32' from assembly 'SixLabors.ImageSharp, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null'.
   at Sense.Led.PixelsFromText.Create(String text)
   at Sense.Led.LedMatrix.ShowMessage(String text)
   at Coreplay2_Local.Program.Main(String[] args) in /usr/src/Program.cs:line 10

**Comment by johannesegger** [comment]
Interesting. I'll see if I can reproduce that soon. Thanks for the report.

**Comment by johannesegger** [comment]
Ok I can reproduce it as soon as I use `SixLabors.ImageSharp 1.0.0-beta0005`. `SenseHatNet` referenced `1.0.0-beta0001` and so just wasn't compatible with it (they had some API changes). Please pull the latest version of the included sample and try running it again. It now uses `SenseHatNet 0.0.3` which references `SixLabors.ImageSharp 1.0.0-beta0005`.

I'll try that as well tomorrow, but I'm not at the RPi at the moment.

**Comment by tgabee2** [comment]
Now at least the ImageSharp exception is gone, but got another one:

Unhandled Exception: System.IO.IOException: Invalid argument
   at System.IO.FileStream.CheckFileCall(Int64 result, Boolean ignoreNotSupported)
   at System.IO.FileStream.Init(FileMode mode, FileShare share)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share)
   at System.IO.File.InternalWriteAllBytes(String path, Byte[] bytes)
   at System.IO.File.WriteAllBytes(String path, Byte[] bytes)
   at Sense.Led.LedMatrix.SetPixels(Pixels pixels)
   at Sense.Led.LedMatrix.ShowMessage(String text)
   at Coreplay2_Local.Program.Main(String[] args) in /usr/src/Program.cs:line 10

The same exception is thrown even if I build/run on the Pi directly, or use Docker.

**Comment by johannesegger** [comment]
Sorry for the delay.

It's [this](https://github.com/johannesegger/SenseHatNet/blob/master/Sense/Led/LedMatrix.cs#L82) call that fails, i.e. you can't write to the frame buffer file.

* Can you check file permissions etc?
* Can you write something to the frame buffer file by hand?
  * Find `/etc/sys/graphics/fb*/name` where the content of the file is `RPi-Sense FB`.
  * Find `/dev/fb*`, where `*` is the same number as in the dir name from above and write something into it.

**Comment by tgabee2** [comment]
> Sorry for the delay.
> 
> It's [this](https://github.com/johannesegger/SenseHatNet/blob/master/Sense/Led/LedMatrix.cs#L82) call that fails, i.e. you can't write to the frame buffer file.
> 
> * Can you check file permissions etc?
> * Can you write something to the frame buffer file by hand?
>   
>   * Find `/etc/sys/graphics/fb*/name` where the content of the file is `RPi-Sense FB`.
>   * Find `/dev/fb*`, where `*` is the same number as in the dir name from above and write something into it.

Hi,

I tried, but there is no /etc/sys path. I use Ubuntu Mate on my Pi3.
I didn't check this buffer on Docker, but the result is the same.

**Comment by johannesegger** [comment]
Omg, I must have been drunk when I wrote the above.

It's not `/etc/sys/graphics/fb*/name`, it's `/sys/class/graphics/fb*/name`. Do you have that?
Also please try [the original Python](https://github.com/RPi-Distro/python-sense-hat) library I copied all the logic from.

**Comment by johannesegger** [comment]
Should be fixed in 0.0.4
